extpreimage/client.go

```diff
@@ -0,0 +1,139 @@
+package extpreimage
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"google.golang.org/grpc"
+	grpcpool "github.com/processout/grpc-go-pool"
+)
+
+// RPC is an interface implemented by the grpc package
+type RPC interface {
+	Dial(host string, opt grpc.DialOption) (*grpc.ClientConn, error)
+	WithInsecure() grpc.DialOption
+	NewClient(*grpcpool.ClientConn) (ExternalPreimageServiceClient)
+}
+
+// grpcRpc exposes the methods from the grpc package that we need
+// this allows us to stub out the grpc methods more easily
+type grpcRpc struct {}
+
+func (r *grpcRpc) Dial(host string, opt grpc.DialOption) (*grpc.ClientConn, error) {
+  return grpc.Dial(host, opt)
+}
+
+func (r *grpcRpc) WithInsecure() grpc.DialOption {
+  return grpc.WithInsecure()
+}
+
+func (r *grpcRpc) NewClient(c *grpcpool.ClientConn) (ExternalPreimageServiceClient) {
+	return NewExternalPreimageServiceClient(c.ClientConn)
+}
+
+// DefaultRPC exposes the default gRPC implementation for consumers
+func DefaultRPC() RPC {
+	return &grpcRpc{}
+}
+
+// Client is the exposed interface for an extpreimage Client
+type Client interface {
+	connect(context.Context) (ExternalPreimageServiceClient, error)
+	Retrieve(*GetPreimageRequest) (*GetPreimageResponse, error)
+	Stop()
+}
+
+// client is a representation of a client of the external preimage
+// service that implements the Client interface
+type client struct {
+	host string
+	conn *grpc.ClientConn
+	client ExternalPreimageServiceClient
+	rpc RPC
+	pool *grpcpool.Pool
+}
+
+// connect creates a new ExternalPreimageServiceClient from the connection pool
+func (c *client) connect(ctx context.Context) (ExternalPreimageServiceClient, error) {
+	conn, err := c.pool.Get(ctx); if err != nil {
+		return nil, err
+	}
+	return c.rpc.NewClient(conn), nil
+}
+
+// Retrieve is a wrapper around the underlying GetPreimage defined
+// in rpc.proto that reduces the stream interface to a single output,
+// since GetPreimage is expected to provide only one response, albeit
+// after a long period of time.
+// Additionally, Retrieve lazily connects to the External Preimage
+// server.
+func (c *client) Retrieve(req *GetPreimageRequest) (*GetPreimageResponse,
+	error) {
```

Comment:
[nit] does this need to be on a new line?

Reply:
LND has a convention of 80 character line length. I just pushed a change to make sure that a couple that were slightly over got brought under the limit.

Reply:
OK sounds good. JIC anyone finds this comment. https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#61-contribution-checklist